File: packages/client/src/pipe/write.rs
use {
	crate::prelude::*,
	futures::{StreamExt as _, stream::BoxStream},
	serde_with::serde_as,
	tangram_http::{Body, response::Ext as _},
	tangram_util::serde::CommaSeparatedString,
};

#[serde_as]
#[derive(Default, Clone, Debug, serde::Deserialize, serde::Serialize)]
pub struct Arg {
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub local: Option<bool>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[serde_as(as = "Option<CommaSeparatedString>")]
	pub remotes: Option<Vec<String>>,
}

impl tg::Client {
	pub async fn write_pipe(
		&self,
		id: &tg::pipe::Id,
		arg: Arg,
		stream: BoxStream<'static, tg::Result<tg::pipe::Event>>,
	) -> tg::Result<()> {
		let method = http::Method::POST;
		let query = serde_urlencoded::to_string(arg)
			.map_err(|source| tg::error!(!source, "failed to serialize the arg"))?;
		let uri = format!("/pipes/{id}/write?{query}");

		// Create the body.
		let body = Body::with_stream(stream.map(move |result| {
			let event = match result {
				Ok(event) => match event {
					tg::pipe::Event::Chunk(bytes) => hyper::body::Frame::data(bytes),
					tg::pipe::Event::End => {
						let mut trailers = http::HeaderMap::new();
						trailers.insert("x-tg-event", http::HeaderValue::from_static("end"));
						hyper::body::Frame::trailers(trailers)
					},
				},
				Err(error) => {
					let mut trailers = http::HeaderMap::new();
					trailers.insert("x-tg-event", http::HeaderValue::from_static("error"));
					let json = serde_json::to_string(&error.to_data_or_id()).unwrap();
					trailers.insert("x-tg-data", http::HeaderValue::from_str(&json).unwrap());
					hyper::body::Frame::trailers(trailers)
				},
			};
			Ok::<_, tg::Error>(event)
		}));

		// Create the request.
		let request = http::request::Builder::default()
			.method(method)
			.uri(uri)
			.header(http::header::ACCEPT, mime::APPLICATION_JSON.to_string())
			.header(
				http::header::CONTENT_TYPE,
				mime::APPLICATION_OCTET_STREAM.to_string(),
			)
			.body(body)
			.unwrap();

		// Send the request.
		let response = self
			.send(request)
			.await
			.map_err(|source| tg::error!(!source, "failed to send the request"))?;
		if !response.status().is_success() {
			let error = response.json().await.map_err(|source| {
				tg::error!(!source, "failed to deserialize the error response")
			})?;
			return Err(error);
		}

		Ok(())
	}
}

File: packages/clients/js/src/client/write.ts
import * as tg from "../index.ts";
import { Body, Request, Uri } from "../http.ts";
import type { Client } from "../client.ts";

export namespace Write {
	export type Arg = {
		checkoutPointers?: boolean;
	};

	export type Output = {
		blob: tg.Referent<tg.Blob.Id>;
	};
}

export async function write(
	client: Client,
	argOrBytes: tg.Write.Arg | string | Uint8Array,
	input?: AsyncIterableIterator<Uint8Array>,
): Promise<tg.Write.Output | tg.Blob.Id> {
	if (typeof argOrBytes === "string" || argOrBytes instanceof Uint8Array) {
		let output = await write(
			client,
			{},
			singleBytes(
				typeof argOrBytes === "string"
					? tg.encoding.utf8.encode(argOrBytes)
					: argOrBytes,
			),
		);
		return (output as tg.Write.Output).blob.node;
	}
	let method = "POST";
	let uri = new Uri({
		path: "/write",
		query: {
			checkout_pointers:
				argOrBytes.checkoutPointers === undefined
					? null
					: argOrBytes.checkoutPointers.toString(),
		},
	});
	let headers = {
		accept: "application/json",
		"content-type": "application/octet-stream",
	};
	tg.assert(input !== undefined);
	let body = new Body(input);
	let request = new Request({
		body,
		method,
		uri,
		headers,
	});
	let response = await client.send(request);
	if (response.status < 200 || response.status >= 300) {
		throw tg.Error.fromData(await response.json<tg.Error.Data>());
	}
	let output = await response.json<{ blob: string }>();
	return {
		blob: tg.Referent.fromDataString(output.blob, (id) => id as tg.Blob.Id),
	};
}

async function* singleBytes(
	bytes: Uint8Array,
): AsyncIterableIterator<Uint8Array> {
	yield bytes;
}
